File: leftShift.c
//Given a singly linked list of size N. The task is to left-shift the linked list by k nodes, where k is a given positive integer smaller than or equal to length of the linked list.

#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

// Create new node
struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

// Insert at end
void insertEnd(struct Node** head, int data) {
    struct Node* newNode = createNode(data);
    if (*head == NULL) {
        *head = newNode;
    } else {
        struct Node* temp = *head;
        while (temp->next != NULL)
            temp = temp->next;
        temp->next = newNode;
    }
}

// Display list
void printList(struct Node* head) {
    while (head != NULL) {
        printf("%d", head->data);
        if (head->next) printf(" -> ");
        head = head->next;
    }
    printf("\n");
}

// Left shift by k
void leftShift(struct Node** head, int k) {
    if (*head == NULL || k == 0)
        return;

    struct Node* temp = *head;
    int count = 1;

    // Traverse to kth node
    while (count < k && temp != NULL) {
        temp = temp->next;
        count++;
    }

    if (temp == NULL)
        return;

    struct Node* kthNode = temp;

    // Traverse to last node
    while (temp->next != NULL)
        temp = temp->next;

    temp->next = *head;      // Connect last to head
    *head = kthNode->next;   // New head
    kthNode->next = NULL;    // Break old link
}

// Main
int main() {
    struct Node* head = NULL;
    insertEnd(&head, 22);
    insertEnd(&head, 44);
    insertEnd(&head, 77);
    insertEnd(&head, 88);
    insertEnd(&head, 99);

    int k = 3;

    printf("Original List:\n");
    printList(head);

    leftShift(&head, k);

    printf("Left Shifted List by %d:\n", k);
    printList(head);

    return 0;
}

File: leftshift.py
#Given a singly linked list of size N. The task is to left-shift the linked list by k nodes, where k is a given positive integer smaller than or equal to length of the linked list. 

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

def insert_end(head, data):
    new_node = Node(data)
    if not head:
        return new_node
    temp = head
    while temp.next:
        temp = temp.next
    temp.next = new_node
    return head

def print_list(head):
    while head:
        print(head.data, end=" -> " if head.next else "\n")
        head = head.next

def left_shift(head, k):
    if not head or k == 0:
        return head

    # Find length
    temp = head
    length = 1
    while temp.next:
        temp = temp.next
        length += 1

    if k >= length:
        return head

    # Traverse to kth node
    temp = head
    for _ in range(k - 1):
        temp = temp.next

    new_head = temp.next
    temp.next = None

    # Find last of new list
    last = new_head
    while last and last.next:
        last = last.next

    if last:
        last.next = head
        return new_head
    else:
        return head

# Driver code
head = None
for val in [22, 44, 77, 88, 99]:
    head = insert_end(head, val)

k = 3
print("Original list:")
print_list(head)

head = left_shift(head, k)

print(f"Left shifted list by {k}:")
print_list(head)
